### PR TITLE
UTY-329: Stop the game in the editor during compilation.

### PR DESCRIPTION
**Pull request by martinzlocha**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Editor hangs frequently if code compiles while the editor is playing and connected to Spatial.

#### Tests
Ran locally and the game restarted after compilation as expected.

#### Documentation
N/A

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
@jessicafalk @samiwh

### CODE REVIEW

File: workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        private static bool stopped;
+
+        [RuntimeInitializeOnLoadMethod]
+        private static void Init()
+        {
+            CompilationPipeline.assemblyCompilationStarted += CompilationPipeline_assemblyCompilationStarted;
+            CompilationPipeline.assemblyCompilationFinished += CompilationPipeline_assemblyCompilationFinished;
+        }
+
+        private static void CompilationPipeline_assemblyCompilationStarted(string obj)
```
Comment (jessicafalk) [inline workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: would rename that to `OnCompilationStarted` and `OnCompilationFinished`

**Comment by samiwh** [comment]
Have this been tested for what happens when you stop play mode while a worker is connecting?

File: workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        private static bool stopped;
+
+        [RuntimeInitializeOnLoadMethod]
+        private static void Init()
+        {
+            CompilationPipeline.assemblyCompilationStarted += OnCompilationStarted;
+            CompilationPipeline.assemblyCompilationFinished += OnCompilationFinished;
+        }
+
+        private static void OnCompilationStarted(string obj)
```
Comment (zeroZshadow) [inline workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: Please give the parameters some names

File: workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            if (EditorApplication.isPlaying)
+            {
+                EditorApplication.isPlaying = false;
+                stopped = true;
+            }
+        }
+
+        private static void OnCompilationFinished(string arg1, CompilerMessage[] arg2)
```
Comment (jared-improbable) [inline workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: What if compilation has failed?

Comment (martinzlocha) [reply workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: It will try to start the game and because there are compiler errors it won't.

**Comment by martinzlocha** [comment]
@samiwh Yes, this does exactly the same thing as if you manually stopped the game while attempting to connect. Haven't seen any weird behaviour so far.

File: workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs (hunk 3 above)
Comment (zeroZshadow) [reply workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: I'd prefer it to not even try to start if compilation failed

Comment (martinzlocha) [reply workers/unity/Assets/Playground/Editor/DisconnectOnCompilation.cs]: Yup, changed that it doesn't try to restart if there are any error messages. But will try if its only warnings.

**Pull request merged**